lib/ruby_lsp/queue.rb

```diff
@@ -0,0 +1,186 @@
+    end
+
+    sig { params(id: T.any(String, Integer)).void }
+    def cancel(id)
+      @mutex.synchronize do
+        # Cancel the job if it's still in the queue
+        @jobs[id]&.cancel
+
+        # Cancel the job if we're in the middle of processing it
+        if @current_job&.request&.dig(:id) == id
+          @worker.raise(Cancelled)
+        end
+      end
+    end
+
+    sig { void }
+    def shutdown
+      # Close the queue so that we can no longer receive items
+      @job_queue.close
+      # Clear any remaining jobs so that the thread can terminate
+      @job_queue.clear
+      # Wait until the thread is finished
+      @worker.join
+    end
+
+    # Executes a request and returns a Queue::Result. No IO should happen in this method, because it can be cancelled in
+    # the middle with a raise
+    sig { params(request: T::Hash[Symbol, T.untyped]).returns(Queue::Result) }
+    def execute(request)
+      response = T.let(nil, T.untyped)
+      error = T.let(nil, T.nilable(StandardError))
+
+      request_time = Benchmark.realtime do
+        response = T.must(@handlers[request[:method]]).action.call(request)
+      rescue Cancelled
+        raise # let the worker handle cancellations
+      rescue StandardError => e
+        error = e
+      end
+
+      Queue::Result.new(response: response, error: error, request_time: request_time)
+    end
+
+    # Finalize a Queue::Result. All IO operations should happen here to avoid any issues with cancelling requests
+    sig do
+      params(
+        result: Result,
+        request: T::Hash[Symbol, T.untyped]
+      ).void
+    end
+    def finalize_request(result, request)
+      error = result.error
+      if error
+        T.must(@handlers[request[:method]]).error_handler&.call(error, request)
+
+        @writer.write(
+          id: request[:id],
+          error: {
+            code: LanguageServer::Protocol::Constant::ErrorCodes::INTERNAL_ERROR,
+            message: result.error.inspect,
+            data: request.to_json,
+          },
+        )
+      elsif result.response != Handler::VOID
+        @writer.write(id: request[:id], result: result.response)
+      end
+
+      @writer.write(method: "telemetry/event", params: telemetry_params(request, result.request_time, result.error))
+    end
+
+    private
+
+    sig { returns(Thread) }
+    def new_worker
+      Thread.new do
+        loop do
+          # Thread::Queue#pop is thread safe and will wait until an item is available
+          job = T.let(@job_queue.pop, T.nilable(Job))
+          # The only time when the job is nil is when the queue is closed and we can then terminate the thread
+          break if job.nil?
+
+          request = job.request
+          @mutex.synchronize do
+            @jobs.delete(request[:id])
+            @current_job = job
+          end
+
+          begin
+            next if job.cancelled
+
+            result = execute(request)
+            finalize_request(result, request)
```

Comment:
Should the finalization happen outside of the cancel handler? It's not cancellable is it?